nx/test/nx/defn/grad_test.exs

```diff
@@ -60,6 +60,51 @@ defmodule Nx.Defn.GradTest do
       assert stop_grad_tuple_meta(Nx.tensor(1), Nx.tensor(1)) == {Nx.tensor(1.0), Nx.tensor(1.0)}
     end
 
+    defn concatenate_grad(t) do
+      grad(t, stop_grad(Nx.sum(Nx.sum(Nx.concatenate([t, t]), axes: [0]))))
+    end
+
+    defn concatenate_grad_power(t) do
+      grad(
+        t,
+        stop_grad(Nx.sum(Nx.sum(Nx.concatenate([Nx.power(t, 2), Nx.power(t, 3)]), axes: [0])))
+      )
+    end
+
+    test "computes concatenate/1 grad for various input shapes" do
```

Comment:
Ok, so after some investigation which involved inserting `print` statements in my local copy of Jax, I managed to get some results which make sense with intermediate `"result"` which is the printed `return ...` from `jax/_src/lax/lax.py -> _concatenate_transpose_rule`, but I'm not sure if these tests should actually be passing or if there's something wrong.

My main concern now is that the final output of `jax.grad` isn't the same as my `grad` calls, and if I remove `stop_grad`, everything goes into an infinite loop.

Reply:
Here is a diff that fixes the loop:

```diff
diff --git a/nx/lib/nx/defn/grad.ex b/nx/lib/nx/defn/grad.ex
index 85b2360..c4fd586 100644
--- a/nx/lib/nx/defn/grad.ex
+++ b/nx/lib/nx/defn/grad.ex
@@ -491,18 +491,17 @@ defmodule Nx.Defn.Grad do
     IO.inspect(starts, label: "starts")
     IO.inspect(limits, label: "limits")
 
-    pre_result = Nx.concatenate(tensors, axis: axis)
-
     starts
     |> Enum.zip(limits)
-    |> Enum.map(fn {start, limit} ->
+    |> Enum.zip(tensors)
+    |> Enum.map(fn {{start, limit}, t} ->
       {start, len} =
         start
         |> Enum.zip(limit)
         |> Enum.map(fn {s, lim} -> {s, lim - s} end)
         |> Enum.unzip()
 
-      {pre_result, Nx.slice(pre_result, start, len)}
+      {t, Nx.slice(ans, start, len)}
     end)
     |> IO.inspect(label: "slices")
     |> grad_pairs(g, cache)
diff --git a/nx/test/nx/defn/grad_test.exs b/nx/test/nx/defn/grad_test.exs
index ebf3229..8775341 100644
--- a/nx/test/nx/defn/grad_test.exs
+++ b/nx/test/nx/defn/grad_test.exs
@@ -61,13 +61,13 @@ defmodule Nx.Defn.GradTest do
     end
 
     defn concatenate_grad(t) do
-      grad(t, stop_grad(Nx.sum(Nx.sum(Nx.concatenate([t, t]), axes: [0]))))
+      grad(t, Nx.sum(Nx.sum(Nx.concatenate([t, t]), axes: [0])))
     end
 
     defn concatenate_grad_power(t) do
       grad(
         t,
-        stop_grad(Nx.sum(Nx.sum(Nx.concatenate([Nx.power(t, 2), Nx.power(t, 3)]), axes: [0])))
+        Nx.sum(Nx.sum(Nx.concatenate([Nx.power(t, 2), Nx.power(t, 3)]), axes: [0]))
       )
     end
 ```

Reply:
Thank you!